segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/rolling/RollingStorage.java

```diff
@@ -321,7 +324,13 @@ public void write(SegmentHandle handle, long offset, InputStream data, int lengt
             int writeLength = (int) Math.min(length - bytesWritten, h.getRollingPolicy().getMaxLength() - last.getLength());
             assert writeLength > 0 : "non-positive write length";
             long chunkOffset = offset + bytesWritten - last.getStartOffset();
-            this.baseStorage.write(h.getActiveChunkHandle(), chunkOffset, data, writeLength);
+
+            // Use a BoundedInputStream to ensure that the underlying storage does not try to read more (or less) data
+            // than we instructed it to. Invoking BoundedInputStream.close() will throw an IOException if baseStorage.write()
+            // has not read all the bytes it was supposed to.
```

Comment:
Here we have BoundedInputStream and the Length of Inputstream (or length to be written) - both are used to control the boundary? Not very clear about the relationship between the two - who's taking priority if the two values conflicting?

Reply:
Actually the requirement is that we want them to be same. That's why we pass same writeLength on line 331 and 332.

Reply:
InputStreams do not have lengths. They have `available()` but that doesn't tell you much.

You need to pass the length to the API to tell it how much to write.

You need to (now) use a BoundedInputStream with a specific bound because some of those Tier 2 APIs do not respect that value. We want to make sure they do not under- or over-read.

Reply:
Then ideally create another method called "writeWithBoundedInputStream" to encapsulate this logic; or at least add a runtime assert here to call out the logic.
A method calling with overlapping input data is error-prone.